core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1525,14 +1516,102 @@ private Scope createScopeForView(
                             false))
                     .collect(toImmutableList());
 
-            analysis.addRelationCoercion(table, outputFields.stream().map(Field::getType).toArray(Type[]::new));
-
-            analyzeFiltersAndMasks(table, name, Optional.empty(), outputFields, session.getIdentity().getUser());
+            if (storageTable.isPresent()) {
+                // use storage table output fields
+                outputFields = analyzeStorageTable(table, outputFields, storageTable.get());
+                analyzeFiltersAndMasks(table, name, storageTable, outputFields, session.getIdentity().getUser());
+                return createAndAssignScope(table, scope, outputFields);
+            }
 
+            analyzeFiltersAndMasks(table, name, storageTable, outputFields, session.getIdentity().getUser());
             outputFields.forEach(field -> analysis.addSourceColumns(field, ImmutableSet.of(new SourceColumn(name, field.getName().orElseThrow()))));
+            analysis.addRelationCoercion(table, outputFields.stream().map(Field::getType).toArray(Type[]::new));
+            analysis.registerNamedQuery(table, query);
             return createAndAssignScope(table, scope, outputFields);
         }
 
+        private List<Field> analyzeStorageTable(Table table, List<Field> viewFields, TableHandle storageTable)
+        {
+            TableSchema tableSchema = metadata.getTableSchema(session, storageTable);
+            Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, storageTable);
+            List<Field> tableFields = analyzeTableOutputFields(table, tableSchema, columnHandles);
+
+            // make sure storage table fields match view fields
+            if (tableFields.size() != viewFields.size()) {
+                throw semanticException(
+                        INVALID_VIEW,
+                        table,
+                        "storage table column count (%s) does not match column count derived from the materialized view query analysis (%s)",
```

Comment:
A non comment, just wondering. Some database systems add a 'row-id' type column to all tables. That will fail this check.

Reply:
good point, I now added a code that skips hidden columns for validation as we do for regular views.